src/NotaryService/NotaryService.cs

```diff
@@ -0,0 +1,401 @@
+                        {
+                            r.MainTx.Witnesses[i] = payload.MainTransaction.Witnesses[i];
+                            r.WitnessInfo[i].NSigsLeft--;
+                        }
+                        break;
+                    case RequestType.MultiSignature:
+                        if (r.WitnessInfo[i].Sigs is null)
+                            r.WitnessInfo[i].Sigs = new Dictionary<ECPoint, byte[]>();
+                        foreach (var pub in r.WitnessInfo[i].Pubs)
+                        {
+                            if (r.WitnessInfo[i].Sigs.ContainsKey(pub))
+                                continue;
+                            if (Crypto.VerifySignature(mainHash, payload.MainTransaction.Witnesses[i].InvocationScript.Skip(2).ToArray(), pub))
+                            {
+                                r.WitnessInfo[i].Sigs[pub] = payload.MainTransaction.Witnesses[i].InvocationScript;
+                                r.WitnessInfo[i].NSigsLeft--;
+                                if (r.WitnessInfo[i].NSigsLeft == 0)
+                                {
+                                    byte[] invScript = Array.Empty<byte>();
+                                    for (int j = 0; j < r.WitnessInfo[i].Pubs.Length; j++)
+                                    {
+                                        if (r.WitnessInfo[i].Sigs.TryGetValue(r.WitnessInfo[i].Pubs[j], out var sig))
+                                            invScript = invScript.Concat(sig).ToArray();
+                                    }
+                                    r.MainTx.Witnesses[i].InvocationScript = invScript;
+                                }
+                                break;
+                            }
+                        }
+                        break;
+                }
+                var currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
+                if (r.IsMainCompleted() && r.MinNotValidBefore > currentHeight)
+                {
+                    Finalize(r.MainTx);
+                    r.IsSent = true;
+                }
+            }
+        }
+
+        private void OnRequestRemoval(NotaryRequest payload)
+        {
+            var r = pendingQueue[payload.MainTransaction.Hash];
+            if (r is null) return;
+            for (int i = 0; i < r.FallbackTxs.Length; i++)
+            {
+                var fb = r.FallbackTxs[i];
+                if (fb.Hash.Equals(payload.FallbackTransaction.Hash))
+                {
+                    var tempList = r.FallbackTxs.ToList();
+                    tempList.RemoveAt(i);
+                    r.FallbackTxs = tempList.ToArray();
+                    break;
+                }
+            }
+            if (r.FallbackTxs.Length == 0) pendingQueue.Remove(r.MainTx.Hash, out _);
+        }
+
+        private void VerifyIncompleteWitnesses(Transaction tx, byte nKeysExpected, out WitnessInfo[] witnessInfos, out string validationErr)
```

Comment:
It would be better if you could have a Error structure such as:

```C#
struct ValidationError{
int value,
string? message
}
```

Reply:
Error type is not fixed